File: poker/games/poker.py
import random
from collections import Counter

class Card:
    """
    トランプのカード1枚を表すクラス。
    """
    SUITS = ["♠", "♥", "♦", "♣"]
    RANKS = ["2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"]
    RANK_VALUES = {r: i for i, r in enumerate(RANKS, 2)}

    def __init__(self, suit, rank):
        self.suit = suit
        self.rank = rank
        self.value = self.RANK_VALUES[rank]

    def __repr__(self):
        return f"{self.suit}{self.rank}"

class Deck:
    """
    52枚のトランプを管理するデッキクラス。
    """
    def __init__(self):
        self.cards = [Card(s, r) for s in Card.SUITS for r in Card.RANKS]
        random.shuffle(self.cards)

    def draw(self):
        return self.cards.pop() if self.cards else None

class PokerHand:
    """
    手札の役判定と強さ比較を行うクラス。
    """
    ROYAL_FLUSH = 9
    STRAIGHT_FLUSH = 8
    FOUR_OF_A_KIND = 7
    FULL_HOUSE = 6
    FLUSH = 5
    STRAIGHT = 4
    THREE_OF_A_KIND = 3
    TWO_PAIR = 2
    ONE_PAIR = 1
    HIGH_CARD = 0

    RANK_NAMES = {
        9: "ロイヤルストレートフラッシュ",
        8: "ストレートフラッシュ",
        7: "フォーカード",
        6: "フルハウス",
        5: "フラッシュ",
        4: "ストレート",
        3: "スリーカード",
        2: "ツーペア",
        1: "ワンペア",
        0: "ハイカード"
    }

    @classmethod
    def evaluate(cls, cards):
        """
        手札(5枚)の役を判定し、(役の強さ, 比較用カード値リスト)を返す。
        """
        values = sorted([c.value for c in cards], reverse=True)
        suits = [c.suit for c in cards]
        counts = Counter(values)
        sorted_counts = counts.most_common()
        
        # フラッシュとストレートの判定
        is_flush = len(set(suits)) == 1
        is_straight = len(set(values)) == 5 and (max(values) - min(values) == 4)
        
        # A,2,3,4,5 のストレート対応
        if not is_straight and set(values) == {14, 2, 3, 4, 5}:
            is_straight = True
            values = [5, 4, 3, 2, 1] # Aを1として扱う
        
        # 役判定
        if is_flush and is_straight and max(values) == 14:
            return cls.ROYAL_FLUSH, values
        if is_flush and is_straight:
            return cls.STRAIGHT_FLUSH, values
        
        if sorted_counts[0][1] == 4:
            return cls.FOUR_OF_A_KIND, [sorted_counts[0][0], sorted_counts[1][0]]
        
        if sorted_counts[0][1] == 3 and sorted_counts[1][1] == 2:
            return cls.FULL_HOUSE, [sorted_counts[0][0], sorted_counts[1][0]]
        
        if is_flush:
            return cls.FLUSH, values
        
        if is_straight:
            return cls.STRAIGHT, values
        
        if sorted_counts[0][1] == 3:
            return cls.THREE_OF_A_KIND, [sorted_counts[0][0]] + sorted([c for c, n in sorted_counts[1:]], reverse=True)
        
        if sorted_counts[0][1] == 2 and sorted_counts[1][1] == 2:
            pairs = sorted([sorted_counts[0][0], sorted_counts[1][0]], reverse=True)
            kicker = sorted_counts[2][0]
            return cls.TWO_PAIR, pairs + [kicker]
        
        if sorted_counts[0][1] == 2:
            pair = sorted_counts[0][0]
            kickers = sorted([c for c, n in sorted_counts[1:]], reverse=True)
            return cls.ONE_PAIR, [pair] + kickers
        
        return cls.HIGH_CARD, values

class PokerGame:
    """
    ポーカーの進行・ルール管理を行うクラス。
    """
    def __init__(self, difficulty="Easy"):
        self.deck = Deck()
        self.difficulty = difficulty
        self.player_hand = []
        self.cpu_hands = []
        self.cpu_count = 1 if difficulty == "Easy" else 3
        
    def start(self):
        """カードを配布する。"""
        self.player_hand = [self.deck.draw() for _ in range(5)]
        self.cpu_hands = [[self.deck.draw() for _ in range(5)] for _ in range(self.cpu_count)]
        
    def exchange_player_cards(self, indices):
        """プレイヤーのカードを交換する。"""
        for i in sorted(indices, reverse=True):
            if 0 <= i < 5:
                self.player_hand[i] = self.deck.draw()
                
    def exchange_cpu_cards(self):
        """
        CPUのカード交換ロジック。
        仕様に基づいた思考を行う。
        """
        for i in range(len(self.cpu_hands)):
            hand = self.cpu_hands[i]
            rank, _ = PokerHand.evaluate(hand)
            
            indices_to_replace = []
            if rank >= PokerHand.FULL_HOUSE:
                # フルハウス以上は交換しない
                pass
            elif rank == PokerHand.THREE_OF_A_KIND:
                # 3枚以外を交換
                counts = Counter([c.value for c in hand])
                three_val = [v for v, c in counts.items() if c == 3][0]
                indices_to_replace = [idx for idx, c in enumerate(hand) if c.value != three_val]
            elif rank == PokerHand.TWO_PAIR:
                # 1枚だけ交換
                counts = Counter([c.value for c in hand])
                pair_vals = [v for v, c in counts.items() if c == 2]
                indices_to_replace = [idx for idx, c in enumerate(hand) if c.value not in pair_vals]
            elif rank == PokerHand.ONE_PAIR:
                # ペア以外を交換
                counts = Counter([c.value for c in hand])
                pair_val = [v for v, c in counts.items() if c == 2][0]
                indices_to_replace = [idx for idx, c in enumerate(hand) if c.value != pair_val]
            else:
                # 役なし: ランダムに2〜3枚交換
                count = random.randint(2, 3)
                indices_to_replace = random.sample(range(5), count)
                
            for idx in indices_to_replace:
                self.cpu_hands[i][idx] = self.deck.draw()

    def determine_winners(self):
        """
        勝敗を判定し、順位(プレイヤーが勝利したか等)を返す。
        """
        p_rank, p_vals = PokerHand.evaluate(self.player_hand)
        p_score = (p_rank, p_vals)
        
        cpu_scores = []
        for h in self.cpu_hands:
            rank, vals = PokerHand.evaluate(h)
            cpu_scores.append((rank, vals))
            
        # プレイヤーと各CPUを比較
        # TrueならプレイヤーがそのCPUより強い
        results = []
        for cs in cpu_scores:
            if p_score > cs:
                results.append(1) # Win
            elif p_score < cs:
                results.append(-1) # Lose
            else:
                results.append(0) # Draw
        
        # 全員に勝利したか、一人にでも負けたか
        if all(r == 1 for r in results):
            return "WIN"
        elif any(r == -1 for r in results):
            return "LOSE"
        else:
            return "DRAW"

File: poker/games/poker.js
/**
 * ポーカーの手札判定とデッキ管理（純粋なロジック）
 */

class Card {
    constructor(suit, rank) {
        this.suit = suit;
        this.rank = rank;
        this.value = this.getValue(rank);
    }
    getValue(rank) {
        const v = { '2': 2, '3': 3, '4': 4, '5': 5, '6': 6, '7': 7, '8': 8, '9': 9, '10': 10, 'J': 11, 'Q': 12, 'K': 13, 'A': 14 };
        return v[rank];
    }
}

class Deck {
    constructor() {
        const suits = ['S', 'H', 'D', 'C'];
        const ranks = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A'];
        this.cards = suits.flatMap(s => ranks.map(r => new Card(s, r)));
        this.shuffle();
    }
    shuffle() {
        for (let i = this.cards.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [this.cards[i], this.cards[j]] = [this.cards[j], this.cards[i]];
        }
    }
    draw() { return this.cards.pop(); }
}

const PokerHand = {
    RANK_NAMES: ["ハイカード", "ワンペア", "ツーペア", "スリーカード", "ストレート", "フラッシュ", "フルハウス", "フォーカード", "ストレートフラッシュ", "ロイヤルストレートフラッシュ"],
    
    evaluate(cards) {
        // カードをインデックス付きで保持
        const indexedCards = cards.map((c, i) => ({ card: c, index: i }));
        const values = indexedCards.map(ic => ic.card.value).sort((a, b) => b - a);
        const suits = indexedCards.map(ic => ic.card.suit);
        
        const counts = {};
        indexedCards.forEach(ic => counts[ic.card.value] = (counts[ic.card.value] || 0) + 1);
        
        const sortedCounts = Object.entries(counts).sort((a, b) => b[1] - a[1] || b[0] - a[0]);
        
        const isFlush = new Set(suits).size === 1;
        let isStraight = new Set(values).size === 5 && (values[0] - values[4] === 4);
        
        // A-5 Straight
        if (!isStraight && values.join(',') === '14,5,4,3,2') {
            isStraight = true;
        }

        let rank = 0;
        let rankIndices = [];
        let compareValues = values;

        if (isFlush && isStraight && values[0] === 14 && values[4] === 10) {
            rank = 9; rankIndices = [0,1,2,3,4];
        } else if (isFlush && isStraight) {
            rank = 8; rankIndices = [0,1,2,3,4];
        } else if (sortedCounts[0][1] === 4) {
            rank = 7;
            const val = Number(sortedCounts[0][0]);
            indexedCards.forEach(ic => { if (ic.card.value === val) rankIndices.push(ic.index); });
            compareValues = [val, Number(sortedCounts[1][0])];
        } else if (sortedCounts[0][1] === 3 && sortedCounts[1][1] === 2) {
            rank = 6;
            rankIndices = [0,1,2,3,4];
            compareValues = [Number(sortedCounts[0][0]), Number(sortedCounts[1][0])];
        } else if (isFlush) {
            rank = 5; rankIndices = [0,1,2,3,4];
        } else if (isStraight) {
            rank = 4; rankIndices = [0,1,2,3,4];
        } else if (sortedCounts[0][1] === 3) {
            rank = 3;
            const val = Number(sortedCounts[0][0]);
            indexedCards.forEach(ic => { if (ic.card.value === val) rankIndices.push(ic.index); });
            compareValues = [val, ...values.filter(v => v !== val)];
        } else if (sortedCounts[0][1] === 2 && sortedCounts[1][1] === 2) {
            rank = 2;
            const v1 = Number(sortedCounts[0][0]);
            const v2 = Number(sortedCounts[1][0]);
            indexedCards.forEach(ic => { if (ic.card.value === v1 || ic.card.value === v2) rankIndices.push(ic.index); });
            compareValues = [v1, v2, ...values.filter(v => v !== v1 && v !== v2)];
        } else if (sortedCounts[0][1] === 2) {
            rank = 1;
            const val = Number(sortedCounts[0][0]);
            indexedCards.forEach(ic => { if (ic.card.value === val) rankIndices.push(ic.index); });
            compareValues = [val, ...values.filter(v => v !== val)];
        }

        return { rank, compareValues, rankIndices };
    }
};
